src/operator/deinput.rs

```diff
@@ -0,0 +1,703 @@
+    /// [`Circuit::add_input_map`](`crate::Circuit::add_input_map`)
+    /// support deletion by key, so the serialized record must match the key
+    /// type of the stream.
+    ///
+    /// The record gets deserialized and pushed to the underlying input stream
+    /// handle as a delete update.
+    ///
+    /// Returns an error if deserialization fails, i.e., the serialized
+    /// representation is corrupted or does not match the value or key
+    /// type of the underlying input stream.
+    ///
+    /// See [`DeZSetHandle`], [`DeSetHandle`], and [`DeMapHandle`]
+    /// documentation for details.
+    fn delete(&mut self, deserializer: &mut dyn ErasedDeserializer) -> Result<(), EError>;
+
+    /// Reserve space for at least `reservation` more updates in the
+    /// internal input buffer.
+    ///
+    /// Reservations are not required but can be used when the number
+    /// of inputs is known ahead of time to reduce reallocations.
+    fn reserve(&mut self, reservation: usize);
+
+    /// Push all buffered updates to the underlying input stream handle.
+    ///
+    /// Flushed updates will be pushed to the stream during the next call
+    /// to [`DBSPHandle::step`](`crate::DBSPHandle::step`).  `flush` can
+    /// be called multiple times between two subsequent `step`s.  Every
+    /// `flush` call adds new updates to the previously flushed updates.
+    ///
+    /// Updates queued after the last `flush` remain buffered in the handle
+    /// until the next `flush` or `clear_buffer` call or until the handle
+    /// is destroyed.
+    fn flush(&mut self);
+
+    /// Clear all buffered updates.
+    ///
+    /// Clears updates pushed to the handle after the last `flush`.
+    /// Flushed updates remain queued at the underlying input handle.
+    // TODO: add another method to invoke `CollectionHandle::clear_input`?
+    fn clear_buffer(&mut self);
+
+    /// Create a new handle connected to the same input stream.
+    ///
+    /// The new handle will use its own input buffer, but shares the
+    /// underlying input stream handle with the original handle.
+    fn fork(&self) -> Box<dyn DeCollectionHandle>;
+}
+
+/// An input handle that wraps a [`CollectionHandle<V, R>`](`CollectionHandle`)
+/// returned by [`Circuit::add_input_zset`](`crate::Circuit::add_input_zset`).
+///
+/// The [`insert`](`Self::insert`) method of this handle deserializes value
+/// `v` type `V` and buffers a `(v, +1)` update for the underlying
+/// `CollectionHandle`.
+///
+/// The [`delete`](`Self::delete`) method of this handle deserializes value
+/// `v` type `V` and buffers a `(v, -1)` update for the underlying
+/// `CollectionHandle`.
+pub struct DeZSetHandle<K, R> {
+    updates: Vec<(K, R)>,
+    handle: CollectionHandle<K, R>,
+}
+
+impl<K, R> DeZSetHandle<K, R> {
+    pub fn new(handle: CollectionHandle<K, R>) -> Self {
+        Self {
+            updates: vec![],
+            handle,
+        }
+    }
+
+    fn clear(&mut self) {
```

Comment:
`clear()` isn't the best name here since no clearing is done, maybe `.retain_buffers()` or something would be better

Reply:
Additionally, the `.shrink_to(MAX_REUSABLE_CAPACITY)` strategy is probably the better one here

Reply:
That's what I did initially, but then I realized that if the buffer is `>MAX_REUSABLE_CAPACITY`, it will get deallocated and a new one will be allocated instead, whereas all I want to do is reuse existing buffer if it's not too huge and free it otherwise.

Reply:
It shouldn't deallocate and reallocate, all it does is make a `realloc()` call if the capacity is larger

Reply:
`realloc` either deallocates and reallocates or leaves the allocation unmodified.  Most memory allocators don't have a way to give part of an allocated memory region back to the allocator.